user_tasks/signals.py

```diff
@@ -0,0 +1,174 @@
+        if parent:
+            parent.increment_total_steps(total_steps)
+
+
+def create_chain_entry(user_id, task_id, task_class, args, kwargs, callbacks, parent=None):
+    """Create and update status records for a new :py:class:`UserTaskMixin` in a Celery chain."""
+    LOGGER.debug(task_class)
+    if issubclass(task_class.__class__, UserTaskMixin):
+        arguments_dict = task_class.arguments_as_dict(*args, **kwargs)
+        name = task_class.generate_name(arguments_dict)
+        total_steps = task_class.calculate_total_steps(arguments_dict)
+        parent_name = kwargs.get('user_task_name', '')
+        if parent is None:
+            # First task in the chain, create a status record for it
+            parent = UserTaskStatus.objects.create(
+                is_container=True, name=parent_name, task_class='celery.chain', task_id=str(uuid4()),
+                total_steps=0, user_id=user_id)
+        UserTaskStatus.objects.create(
+            name=name, parent=parent, task_class=task_class, task_id=task_id, total_steps=total_steps, user_id=user_id)
+        parent.increment_total_steps(total_steps)
+        if parent_name and not parent.name:
+            parent.set_name(parent_name)
+    for callback in callbacks:
+        links = callback.options.get('link', [])
+        callback_class = import_string(callback.task)
+        create_chain_entry(user_id, callback.id, callback_class, callback.args, callback.kwargs, links, parent=parent)
+
+
+def create_chord_entry(task_id, task_class, message_body, user_id):
+    """Create and update status records for a new :py:class:`UserTaskMixin` in a Celery chord."""
+    args = message_body['args']
+    kwargs = message_body['kwargs']
+    arguments_dict = task_class.arguments_as_dict(*args, **kwargs)
+    name = task_class.generate_name(arguments_dict)
+    total_steps = task_class.calculate_total_steps(arguments_dict)
+    parent_name = kwargs.get('user_task_name', '')
+    chord_data = message_body['chord']
+    group_id = message_body['taskset']
+    group, created = UserTaskStatus.objects.get_or_create(
+        task_id=group_id, defaults={'is_container': True, 'name': parent_name, 'task_class': 'celery.group',
+                                    'total_steps': total_steps, 'user_id': user_id})
+    if created:
+        # Also create a status for the chord as a whole
+        chord = UserTaskStatus.objects.create(
+            is_container=True, name=parent_name, task_class='celery.chord', task_id=str(uuid4()),
+            total_steps=total_steps, user_id=user_id)
+        group.parent = chord
+        group.save(update_fields={'parent', 'modified'})
+    else:
+        chord = None
+        group.increment_total_steps(total_steps)
+        if parent_name and not group.name:
+            group.set_name(parent_name)
+    UserTaskStatus.objects.create(
+        name=name, parent=group, task_class=task_class, task_id=task_id, total_steps=total_steps, user_id=user_id)
```

Comment:
I suppose it wouldn't hurt, since this code is run before the task is announced to any celery workers which could trigger any deadlocks working with the same records.